<a href="https://colab.research.google.com/github/venkateshblks/NLP-Hugging-Face/blob/main/NLP_hugging_face.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
# !pip install transformers

# Transformer Models

## sentiment-analysis

In [ ]:
from transformers import pipeline
cl=pipeline('sentiment-analysis')

No model was supplied, defaulted to distilbert/distilbert-base-uncased-finetuned-sst-2-english and revision 714eb0f (https://huggingface.co/distilbert/distilbert-base-uncased-finetuned-sst-2-english).
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json:   0%|          | 0.00/629 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/268M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

Device set to use cpu


In [ ]:
cl('What a beautiful day! The sun is shining and the weather is perfect')

[{'label': 'POSITIVE', 'score': 0.9998867511749268}]

In [ ]:
cl(['I hated the movie. It was a complete waste of time',
    'The food was delicious, but the service was extremely slow',
    'This restaurant has the best food in town'])

[{'label': 'NEGATIVE', 'score': 0.9998189806938171},
 {'label': 'NEGATIVE', 'score': 0.9946165680885315},
 {'label': 'POSITIVE', 'score': 0.9998156428337097}]

## Zero-shot classification

**it allows you to specify which labels to use for the classification**

In [ ]:
cl=pipeline('zero-shot-classification')

No model was supplied, defaulted to facebook/bart-large-mnli and revision d7645e1 (https://huggingface.co/facebook/bart-large-mnli).
Using a pipeline without specifying a model name and revision in production is not recommended.
Device set to use cpu


In [ ]:
cl

In [ ]:
cl("The new curriculum introduced in schools aims to foster critical thinking and problem-solving skills",candidate_labels=["education", "politics", "business"],)

{'sequence': 'The new curriculum introduced in schools aims to foster critical thinking and problem-solving skills',
 'labels': ['education', 'business', 'politics'],
 'scores': [0.9582398533821106, 0.02940666303038597, 0.012353500351309776]}

In [ ]:
cl(["The new curriculum introduced in schools aims to foster critical thinking and problem-solving skills",
    "The company's quarterly earnings exceeded expectations, boosting investor confidence",
    ],candidate_labels=["education", "politics", "business"],)

[{'sequence': 'The new curriculum introduced in schools aims to foster critical thinking and problem-solving skills',
  'labels': ['education', 'business', 'politics'],
  'scores': [0.9582398533821106, 0.02940666303038597, 0.012353500351309776]},
 {'sequence': "The company's quarterly earnings exceeded expectations, boosting investor confidence",
  'labels': ['business', 'education', 'politics'],
  'scores': [0.9799111485481262, 0.0118860537186265, 0.008202859200537205]}]

## Text generation

In [ ]:
generator = pipeline("text-generation")

No model was supplied, defaulted to openai-community/gpt2 and revision 607a30d (https://huggingface.co/openai-community/gpt2).
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json:   0%|          | 0.00/665 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/548M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/124 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/1.04M [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]

Device set to use cpu


In [ ]:
generator("Once upon a time, in a land far away, there was a kingdom ruled by",num_return_sequences=3)

Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


[{'generated_text': 'Once upon a time, in a land far away, there was a kingdom ruled by the gods of darkness. It was called the White Tower, for as the moon was so often called, the tower was called its name. And the city of the'},
 {'generated_text': 'Once upon a time, in a land far away, there was a kingdom ruled by mighty men. But if they had given up their power, who would have dreamed that such a country would someday be under the dominion of kings? It was from this'},
 {'generated_text': 'Once upon a time, in a land far away, there was a kingdom ruled by three kings; from the three kings, they were called Great Kings. Thereupon they came to be called, or were called Kinglings, or Great Kings under the'}]

**choose a particular model from the Hub to use in a pipeline for a specific task**

In [ ]:
generator = pipeline("text-generation", model="distilgpt2")

config.json:   0%|          | 0.00/762 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/353M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/124 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/1.04M [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]

Device set to use cpu


In [ ]:
generator("Once upon a time, in a land far away, there was a kingdom ruled by",num_return_sequences=3)

Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


[{'generated_text': 'Once upon a time, in a land far away, there was a kingdom ruled by King Mormont: the kingdom of his men.\n\n\n\nIn the aftermath of the Battle of the Great Stone, King Ayrne sent warriors back'},
 {'generated_text': "Once upon a time, in a land far away, there was a kingdom ruled by king of the Haji'i.\n\n\nOn one of those days in Jazahu (in Diklahu, the capital of Jaz"},
 {'generated_text': 'Once upon a time, in a land far away, there was a kingdom ruled by one man who was going to have the power that is only to be brought to their knees. With this, there were more. It came down to two things:'}]

## Mask filling

In [ ]:
unmasker = pipeline("fill-mask")

No model was supplied, defaulted to distilbert/distilroberta-base and revision fb53ab8 (https://huggingface.co/distilbert/distilroberta-base).
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json:   0%|          | 0.00/480 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/331M [00:00<?, ?B/s]

Some weights of the model checkpoint at distilbert/distilroberta-base were not used when initializing RobertaForMaskedLM: ['roberta.pooler.dense.bias', 'roberta.pooler.dense.weight']
- This IS expected if you are initializing RobertaForMaskedLM from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing RobertaForMaskedLM from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).


tokenizer_config.json:   0%|          | 0.00/25.0 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/899k [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]

Device set to use cpu


In [ ]:
unmasker("The quick brown fox jumped over the <mask> dog")

[{'score': 0.38676905632019043,
  'token': 35828,
  'token_str': ' barking',
  'sequence': 'The quick brown fox jumped over the barking dog'},
 {'score': 0.06970981508493423,
  'token': 21480,
  'token_str': ' stray',
  'sequence': 'The quick brown fox jumped over the stray dog'},
 {'score': 0.03620748966932297,
  'token': 37747,
  'token_str': ' startled',
  'sequence': 'The quick brown fox jumped over the startled dog'},
 {'score': 0.022873898968100548,
  'token': 8416,
  'token_str': ' sleeping',
  'sequence': 'The quick brown fox jumped over the sleeping dog'},
 {'score': 0.017269104719161987,
  'token': 26851,
  'token_str': ' frightened',
  'sequence': 'The quick brown fox jumped over the frightened dog'}]

In [ ]:
unmasker( "Einstein is known for his theory of <mask>",top_k=3,)

[{'score': 0.9341339468955994,
  'token': 44999,
  'token_str': ' relativity',
  'sequence': 'Einstein is known for his theory of relativity'},
 {'score': 0.02886814810335636,
  'token': 10795,
  'token_str': ' evolution',
  'sequence': 'Einstein is known for his theory of evolution'},
 {'score': 0.0072267684154212475,
  'token': 17244,
  'token_str': ' gravity',
  'sequence': 'Einstein is known for his theory of gravity'}]

In [ ]:
unmasker("This is a simple <mask>. " + "...with a large amount of repeated text appended. " * 5,tokenizer_kwargs={"padding":True},top_k=3)

[{'score': 0.17382566630840302,
  'token': 1246,
  'token_str': ' example',
  'sequence': 'This is a simple example. ...with a large amount of repeated text appended. ...with a large amount of repeated text appended. ...with a large amount of repeated text appended. ...with a large amount of repeated text appended. ...with a large amount of repeated text appended. '},
 {'score': 0.04235488176345825,
  'token': 19587,
  'token_str': ' equation',
  'sequence': 'This is a simple equation. ...with a large amount of repeated text appended. ...with a large amount of repeated text appended. ...with a large amount of repeated text appended. ...with a large amount of repeated text appended. ...with a large amount of repeated text appended. '},
 {'score': 0.04157842695713043,
  'token': 21586,
  'token_str': ' calculation',
  'sequence': 'This is a simple calculation. ...with a large amount of repeated text appended. ...with a large amount of repeated text appended. ...with a large amount of rep

## Named entity recognition

In [ ]:
ner = pipeline("ner", grouped_entities=True)

No model was supplied, defaulted to dbmdz/bert-large-cased-finetuned-conll03-english and revision 4c53496 (https://huggingface.co/dbmdz/bert-large-cased-finetuned-conll03-english).
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json:   0%|          | 0.00/998 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/1.33G [00:00<?, ?B/s]

Some weights of the model checkpoint at dbmdz/bert-large-cased-finetuned-conll03-english were not used when initializing BertForTokenClassification: ['bert.pooler.dense.bias', 'bert.pooler.dense.weight']
- This IS expected if you are initializing BertForTokenClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertForTokenClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).


tokenizer_config.json:   0%|          | 0.00/60.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/213k [00:00<?, ?B/s]

Device set to use cpu
/usr/local/lib/python3.10/dist-packages/transformers/pipelines/token_classification.py:170: UserWarning: `grouped_entities` is deprecated and will be removed in version v5.0.0, defaulted to `aggregation_strategy="simple"` instead.
  warnings.warn(


In [ ]:
ner("My name is venkatesh and I work at Google  in india.")

[{'entity_group': 'PER',
  'score': 0.9886713,
  'word': 'venkatesh',
  'start': 11,
  'end': 20},
 {'entity_group': 'ORG',
  'score': 0.99875355,
  'word': 'Google',
  'start': 35,
  'end': 41},
 {'entity_group': 'LOC',
  'score': 0.7703972,
  'word': 'india',
  'start': 46,
  'end': 51}]

## Question answering

In [ ]:
question_answerer = pipeline("question-answering")

No model was supplied, defaulted to distilbert/distilbert-base-cased-distilled-squad and revision 564e9b5 (https://huggingface.co/distilbert/distilbert-base-cased-distilled-squad).
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json:   0%|          | 0.00/473 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/261M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/49.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/213k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/436k [00:00<?, ?B/s]

Device set to use cpu


In [ ]:
context = """
Albert Einstein was a theoretical physicist who developed the theory of relativity, one of the two pillars of modern physics
(which includes quantum mechanics). His work is also known for its influence on the philosophy of science. He was awarded the
Nobel Prize in Physics in 1921 for his explanation of the photoelectric effect, a pivotal step in the development of quantum theory.
Einstein's contributions to science have made him one of the most well-known and influential scientists in history.
"""

# Ask a question related to the context
question = "What did Albert Einstein win the Nobel Prize for?"

question_answerer(question=question, context=context)


{'score': 0.36903250217437744,
 'start': 293,
 'end': 332,
 'answer': 'explanation of the photoelectric effect'}

In [ ]:
context = """
The capital of France is Paris. Paris is known for its beautiful landmarks such as the Eiffel Tower and the Louvre Museum.
Many tourists visit Paris every year to experience its culture, food, and art.
"""
question = "What is the capital of France?"
question_answerer(question=question, context=context)

{'score': 0.9469671249389648, 'start': 26, 'end': 31, 'answer': 'Paris'}

## Summarization

In [ ]:
summarizer = pipeline("summarization")

No model was supplied, defaulted to sshleifer/distilbart-cnn-12-6 and revision a4f8f3e (https://huggingface.co/sshleifer/distilbart-cnn-12-6).
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json:   0%|          | 0.00/1.80k [00:00<?, ?B/s]

pytorch_model.bin:   0%|          | 0.00/1.22G [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/899k [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

Device set to use cpu


In [ ]:
s="""Investors rushing to capitalize on artificial intelligence have focused on the technology—the capabilities of new model the potential of generative tools,
 and the scale of processing power to sustain it all.
 What too many ignore is the evolving legal structure surrounding the technology,
 which will ultimately shape the economics of AI.  The core question is: Who controls the value that AI produces?
 The answer depends on whether AI companies must compensate rights holders for using their data to train AI models and
 whether AI creations can themselves enjoy copyright or patent protections.

 The current landscape of AI law is rife with uncertainty. Getty Images and individual artists have challenged AI companies over their use of copyrighted material in training data sets.
 How these cases are decided will determine whether AI developers can harvest publicly available data or must license the content used to train their models.
 Should courts decide in favor of rights holders,AI companies will face increased costs that could reduce profit margins and put many current valuations into question.
 Investors shouldn’t underestimate the risks.
 """
summarizer(s)

[{'summary_text': ' Investors rushing to capitalize on artificial intelligence have focused on the technology . But too many ignore the evolving legal structure surrounding the technology, which will ultimately shape the economics of AI . The core question is: Who controls the value that AI produces? The answer depends on whether AI companies must compensate rights holders for using their data .'}]

In [ ]:
summarizer(s, do_sample=True)

[{'summary_text': ' Investors rushing to capitalize on artificial intelligence have focused on the technology . What too many ignore is the evolving legal structure surrounding the technology, which will ultimately shape the economics of AI . The answer depends on whether AI companies must compensate rights holders for using their data to train AI models .'}]

## Translation

In [ ]:
translator1=pipeline("translation", model="Helsinki-NLP/opus-mt-en-fr")  # English to French

Device set to use cpu


In [ ]:
translator1("Hello, how are you today?")

[{'translation_text': "Bonjour, comment allez-vous aujourd'hui ?"}]

In [ ]:
translator = pipeline("translation", model="Helsinki-NLP/opus-mt-fr-en")

Device set to use cpu


In [ ]:
translator("Bonjour, comment allez-vous aujourd'hui ?")  #French to English

[{'translation_text': 'Hello, how are you today?'}]

# Using Transformers

## Behind the pipeline (PyTorch)

<img src="https://huggingface.co/datasets/huggingface-course/documentation-images/resolve/main/en/chapter2/full_nlp_pipeline.svg" />

In [ ]:
cl=pipeline('sentiment-analysis')
raw_text=['I hated the movie. It was a complete waste of time',
    'The food was delicious, but the service was extremely slow',
    'This restaurant has the best food in town']
cl(raw_text)

No model was supplied, defaulted to distilbert/distilbert-base-uncased-finetuned-sst-2-english and revision 714eb0f (https://huggingface.co/distilbert/distilbert-base-uncased-finetuned-sst-2-english).
Using a pipeline without specifying a model name and revision in production is not recommended.
Device set to use cpu


[{'label': 'NEGATIVE', 'score': 0.9998189806938171},
 {'label': 'NEGATIVE', 'score': 0.9946165680885315},
 {'label': 'POSITIVE', 'score': 0.9998156428337097}]

In [ ]:
from transformers import AutoTokenizer
checkpoint = "distilbert-base-uncased-finetuned-sst-2-english"
tokenizer = AutoTokenizer.from_pretrained(checkpoint)

In [ ]:
input_ids= tokenizer(raw_text, padding=True, truncation=True, return_tensors="pt")
input_ids

{'input_ids': tensor([[  101,  1045,  6283,  1996,  3185,  1012,  2009,  2001,  1037,  3143,
          5949,  1997,  2051,   102],
        [  101,  1996,  2833,  2001, 12090,  1010,  2021,  1996,  2326,  2001,
          5186,  4030,   102,     0],
        [  101,  2023,  4825,  2038,  1996,  2190,  2833,  1999,  2237,   102,
             0,     0,     0,     0]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
        [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0],
        [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0]])}

In [ ]:
from transformers import AutoModelForSequenceClassification
model = AutoModelForSequenceClassification.from_pretrained(checkpoint)
outputs = model(**input_ids)

In [ ]:
outputs

SequenceClassifierOutput(loss=None, logits=tensor([[ 4.7628, -3.8538],
        [ 2.8316, -2.3874],
        [-4.1376,  4.4607]], grad_fn=<AddmmBackward0>), hidden_states=None, attentions=None)

In [ ]:
outputs.logits

tensor([[ 4.7628, -3.8538],
        [ 2.8316, -2.3874],
        [-4.1376,  4.4607]], grad_fn=<AddmmBackward0>)

In [ ]:
import torch

predictions = torch.nn.functional.softmax(outputs.logits, dim=-1)
predictions

tensor([[9.9982e-01, 1.8104e-04],
        [9.9462e-01, 5.3834e-03],
        [1.8440e-04, 9.9982e-01]], grad_fn=<SoftmaxBackward0>)

**First sentence:** NEGATIVE: 0.9998, POSITIVE: 0.00018

**Second sentence:** NEGATIVE: 0.9946, POSITIVE: 0.0053

**third sentence:** NEGATIVE: 0.000184, POSITIVE: 0.9998

In [ ]:
torch.argmax(outputs.logits,dim=-1)

tensor([0, 0, 1])

In [ ]:
model.config.id2label

{0: 'NEGATIVE', 1: 'POSITIVE'}

## Tokenizers

In [7]:
from transformers import BertTokenizer

tokenizer = BertTokenizer.from_pretrained("bert-base-cased")

tokenizer_config.json:   0%|          | 0.00/49.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/213k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/436k [00:00<?, ?B/s]

In [14]:
tokenizer("Using a Transformer network is simple")


{'input_ids': [101, 7993, 170, 13809, 23763, 2443, 1110, 3014, 102], 'token_type_ids': [0, 0, 0, 0, 0, 0, 0, 0, 0], 'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1]}

In [9]:
from transformers import AutoTokenizer

tokenizer = AutoTokenizer.from_pretrained("bert-base-cased")

In [31]:
tokenizer("Using a Transformer network is simple")

{'input_ids': [101, 7993, 170, 13809, 23763, 2443, 1110, 3014, 102], 'token_type_ids': [0, 0, 0, 0, 0, 0, 0, 0, 0], 'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1]}

In [29]:
tokens=tokenizer.tokenize('Using a Transformer network is simple')
tokens

['Using', 'a', 'Trans', '##former', 'network', 'is', 'simple']

In [20]:
ids=tokenizer.convert_tokens_to_ids(tokens)
ids

[7993, 170, 13809, 23763, 2443, 1110, 3014]

In [21]:
tokenizer.decode(ids)

'Using a Transformer network is simple'

In [28]:
tokenizer.save_pretrained("Saving_a_tokenizer")

('Saving_a_tokenizer/tokenizer_config.json',
 'Saving_a_tokenizer/special_tokens_map.json',
 'Saving_a_tokenizer/vocab.txt',
 'Saving_a_tokenizer/added_tokens.json',
 'Saving_a_tokenizer/tokenizer.json')

## Handling Multiple sequences

In [84]:
import torch
from transformers import AutoTokenizer, AutoModelForSequenceClassification

checkpoint = "distilbert-base-uncased-finetuned-sst-2-english"
tokenizer = AutoTokenizer.from_pretrained(checkpoint)
model = AutoModelForSequenceClassification.from_pretrained(checkpoint)

In [85]:

sequence = "This restaurant has the best food in town"

tokens = tokenizer.tokenize(sequence)
ids = tokenizer.convert_tokens_to_ids(tokens)
print(tokens)
ids

['this', 'restaurant', 'has', 'the', 'best', 'food', 'in', 'town']


[2023, 4825, 2038, 1996, 2190, 2833, 1999, 2237]

In [86]:
input_ids = torch.tensor([ids])
input_ids

tensor([[2023, 4825, 2038, 1996, 2190, 2833, 1999, 2237]])

In [87]:
output = model(input_ids)
# print(output)
output.logits

tensor([[-3.3032,  3.5646]], grad_fn=<AddmmBackward0>)

In [77]:
torch.nn.functional.softmax(output.logits, dim=-1)

tensor([[0.0010, 0.9990]], grad_fn=<SoftmaxBackward0>)

**When you’re trying to batch together two (or more) sentences, they might be of different lengths. If you’ve ever worked with tensors before, you know that they need to be of rectangular shape, so you won’t be able to convert the list of input IDs into a tensor directly**

In [105]:
batched_ids = [
    [200, 200, 200],    #  ValueError: expected sequence of length 3 at dim 1 (got 2)
    [200, 200]
]

# we’ll use padding to make our tensors have a rectangular shape.
# Padding makes sure all our sentences have the same length by adding a special word called the padding token to the sentences with fewer values
batched_ids = [
    [200, 200, 200],
    [200, 200,0]
]
print(model(torch.tensor(batched_ids)).logits)
batched_ids = [
    [200, 200, 200],
    [200, 200,100]
]
model(torch.tensor(batched_ids)).logits

tensor([[ 1.5694, -1.3895],
        [ 1.3374, -1.2163]], grad_fn=<AddmmBackward0>)


tensor([[ 1.5694, -1.3895],
        [ 0.9907, -0.9139]], grad_fn=<AddmmBackward0>)

 To get the same result when passing individual sentences of different lengths through the model or when passing a batch with the same sentences and padding applied, we need to tell those attention layers to ignore the padding tokens. This is done by using an attention mask.

In [104]:
batched_ids = [
    [200, 200, 200],
    [200, 200,0]
]
attention_mask = [
    [1, 1, 1],
    [1, 1, 0],
]
print(model(torch.tensor(batched_ids),torch.tensor(attention_mask)).logits)
batched_ids = [
    [200, 200, 200],
    [200, 200,200]
]
attention_mask = [
    [1, 1, 1],
    [1, 1, 0],
]
model(torch.tensor(batched_ids),torch.tensor(attention_mask)).logits

tensor([[ 1.5694, -1.3895],
        [ 0.5803, -0.4125]], grad_fn=<AddmmBackward0>)


tensor([[ 1.5694, -1.3895],
        [ 0.5803, -0.4125]], grad_fn=<AddmmBackward0>)

Now we get the same logits

**Attention masks are tensors with the exact same shape as the input IDs tensor, filled with 0s and 1s: 1s indicate the corresponding tokens should be attended to, and 0s indicate the corresponding tokens should not be attended to (i.e., they should be ignored by the attention layers of the model)**

In [79]:
tokenizer.pad_token_id

0

## Putting it all together

In [107]:
from transformers import AutoTokenizer

checkpoint = "distilbert-base-uncased-finetuned-sst-2-english"
tokenizer = AutoTokenizer.from_pretrained(checkpoint)

sequence = "I've been waiting for a HuggingFace course my whole life."

model_inputs = tokenizer(sequence)
model_inputs

{'input_ids': [101, 1045, 1005, 2310, 2042, 3403, 2005, 1037, 17662, 12172, 2607, 2026, 2878, 2166, 1012, 102], 'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]}

In [113]:
sequences = ["I've been waiting for a HuggingFace course my whole life.", "So have I!"]

model_inputs = tokenizer(sequences)
model_inputs

{'input_ids': [[101, 1045, 1005, 2310, 2042, 3403, 2005, 1037, 17662, 12172, 2607, 2026, 2878, 2166, 1012, 102], [101, 2061, 2031, 1045, 999, 102]], 'attention_mask': [[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1], [1, 1, 1, 1, 1, 1]]}

In [123]:
# Will pad the sequences up to the maximum sequence length
model_inputs = tokenizer(sequences, padding="longest")
print(len(model_inputs.input_ids[1]))
print(model_inputs)
# Will pad the sequences up to the model max length
# (512 for BERT or DistilBERT)
model_inputs = tokenizer(sequences, padding="max_length")
print(len(model_inputs.input_ids[1]))
print(model_inputs)

# Will pad the sequences up to the specified max length
model_inputs = tokenizer(sequences, padding="max_length", max_length=8)
print(len(model_inputs.input_ids[1]))
model_inputs

16
{'input_ids': [[101, 1045, 1005, 2310, 2042, 3403, 2005, 1037, 17662, 12172, 2607, 2026, 2878, 2166, 1012, 102], [101, 2061, 2031, 1045, 999, 102, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]], 'attention_mask': [[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1], [1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]]}
512
{'input_ids': [[101, 1045, 1005, 2310, 2042, 3403, 2005, 1037, 17662, 12172, 2607, 2026, 2878, 2166, 1012, 102, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 

{'input_ids': [[101, 1045, 1005, 2310, 2042, 3403, 2005, 1037, 17662, 12172, 2607, 2026, 2878, 2166, 1012, 102], [101, 2061, 2031, 1045, 999, 102, 0, 0]], 'attention_mask': [[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1], [1, 1, 1, 1, 1, 1, 0, 0]]}

In [129]:
sequences = ["I've been waiting for a HuggingFace course my whole life.", "So have I!"]

# Will truncate the sequences that are longer than the model max length
# (512 for BERT or DistilBERT)
# If the sequence length is less than or equal to the max length, truncation will not occur.
model_inputs = tokenizer(sequences, truncation=True)
print(len(model_inputs.input_ids[0]))
print(model_inputs)

# Will truncate the sequences that are longer than the specified max length
model_inputs = tokenizer(sequences, max_length=8, truncation=True)
print(len(model_inputs.input_ids[0]))
model_inputs

16
{'input_ids': [[101, 1045, 1005, 2310, 2042, 3403, 2005, 1037, 17662, 12172, 2607, 2026, 2878, 2166, 1012, 102], [101, 2061, 2031, 1045, 999, 102]], 'attention_mask': [[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1], [1, 1, 1, 1, 1, 1]]}
8


{'input_ids': [[101, 1045, 1005, 2310, 2042, 3403, 2005, 102], [101, 2061, 2031, 1045, 999, 102]], 'attention_mask': [[1, 1, 1, 1, 1, 1, 1, 1], [1, 1, 1, 1, 1, 1]]}

In [135]:
sequences = ["I've been waiting for a HuggingFace course my whole life.", "So have I!"]

# Returns PyTorch tensors
model_inputs = tokenizer(sequences, padding=True, return_tensors="pt")
print('PyTorch tensors')
print(model_inputs)
print("-"*100)

# Returns TensorFlow tensors
model_inputs = tokenizer(sequences, padding=True, return_tensors="tf")
print('TensorFlow tensors')
print(model_inputs)
print("-"*100)

# Returns NumPy arrays
model_inputs = tokenizer(sequences, padding=True, return_tensors="np")
print('NumPy tensors')
model_inputs

PyTorch tensors
{'input_ids': tensor([[  101,  1045,  1005,  2310,  2042,  3403,  2005,  1037, 17662, 12172,
          2607,  2026,  2878,  2166,  1012,   102],
        [  101,  2061,  2031,  1045,   999,   102,     0,     0,     0,     0,
             0,     0,     0,     0,     0,     0]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
        [1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]])}
----------------------------------------------------------------------------------------------------
TensorFlow tensors
{'input_ids': <tf.Tensor: shape=(2, 16), dtype=int32, numpy=
array([[  101,  1045,  1005,  2310,  2042,  3403,  2005,  1037, 17662,
        12172,  2607,  2026,  2878,  2166,  1012,   102],
       [  101,  2061,  2031,  1045,   999,   102,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0]], dtype=int32)>, 'attention_mask': <tf.Tensor: shape=(2, 16), dtype=int32, numpy=
array([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 

{'input_ids': array([[  101,  1045,  1005,  2310,  2042,  3403,  2005,  1037, 17662,
        12172,  2607,  2026,  2878,  2166,  1012,   102],
       [  101,  2061,  2031,  1045,   999,   102,     0,     0,     0,
            0,     0,     0,     0,     0,     0,     0]]), 'attention_mask': array([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
       [1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]])}

In [136]:
sequence = "I've been waiting for a HuggingFace course my whole life."

model_inputs = tokenizer(sequence)
print(model_inputs["input_ids"])

tokens = tokenizer.tokenize(sequence)
ids = tokenizer.convert_tokens_to_ids(tokens)
print(ids)

[101, 1045, 1005, 2310, 2042, 3403, 2005, 1037, 17662, 12172, 2607, 2026, 2878, 2166, 1012, 102]
[1045, 1005, 2310, 2042, 3403, 2005, 1037, 17662, 12172, 2607, 2026, 2878, 2166, 1012]


In [137]:
print(tokenizer.decode(model_inputs["input_ids"]))
print(tokenizer.decode(ids))

[CLS] i've been waiting for a huggingface course my whole life. [SEP]
i've been waiting for a huggingface course my whole life.


In [143]:
sequences = ["I've been waiting for a HuggingFace course my whole life.", "So have I!"]

tokens = tokenizer(sequences, padding=True, truncation=True, return_tensors="pt")
print(tokens)
output = model(**tokens)
output.logits

{'input_ids': tensor([[  101,  1045,  1005,  2310,  2042,  3403,  2005,  1037, 17662, 12172,
          2607,  2026,  2878,  2166,  1012,   102],
        [  101,  2061,  2031,  1045,   999,   102,     0,     0,     0,     0,
             0,     0,     0,     0,     0,     0]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
        [1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]])}


tensor([[-1.5607,  1.6123],
        [-3.6183,  3.9137]], grad_fn=<AddmmBackward0>)

In [144]:
output = model(tokens.input_ids)
output.logits

tensor([[-1.5607,  1.6123],
        [-0.2816,  0.4363]], grad_fn=<AddmmBackward0>)